### PR TITLE
feat(meta): initiate static compaction groups

### PR DESCRIPTION
**Pull request by zwang28**
## What's changed and what's your intention?
Changes:

- Create 2 static compaction groups in meta node. Compaction group is maintained via 2 mappings in HummockManager and CompactionGroupManager:

  - `compaction_group_id ` -> `CompactStatus`. It tells which SSTs are being compacted. It will be updated per compaction, which is relatively frequent.
  - `compaction_group_id ` -> `CompactionGroup`. It stores registered prefixes, SSTs configuration and compaction strategy configuration. It is updated less frequently than the former mapping.

- Add a prost `CompactionConfig` to persist config.

- Move prost code in compaction module to prost_type.rs.

## Checklist

- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/2065

### CODE REVIEW

File: src/meta/src/hummock/compaction/mod.rs [hunk 1]
```diff
@@ -90,92 +86,57 @@ pub struct SearchResult {
...
         &mut self,
         levels: &[Level],
         task_id: HummockCompactionTaskId,
+        compaction_group_id: CompactionGroupId,
     ) -> Option<CompactTask> {
         // When we compact the files, we must make the result of compaction meet the following
         // conditions, for any user key, the epoch of it in the file existing in the lower
         // layer must be larger.
 
+        // TODO #2065: filter levels by compaction_group_id so that only touch SSTs within given
```
Comment (soundOfDestiny) [inline src/meta/src/hummock/compaction/mod.rs:137]: does it mean that different levels will have different compaction_group_id?

Comment (zwang28) [reply src/meta/src/hummock/compaction/mod.rs:137]: No. Let me rephrase the comment: ~filter levels by compaction_group_id~ filter out SSTs that doesn't belong to `compaction_group_id`.

Comment (zwang28) [reply src/meta/src/hummock/compaction/mod.rs:137]: By design we've ensured one SST only belongs to one compaction group. So we can decide a SST's belonging based on `SstableInfo`.

File: src/meta/src/hummock/hummock_manager.rs [hunk 2]
```diff
@@ -75,9 +74,9 @@ pub struct HummockManager<S: MetaStore> {
 pub type HummockManagerRef<S> = Arc<HummockManager<S>>;
 
 struct Compaction {
-    // TODO: `compact_status` will be managed per `compaction_groups` soon
-    compact_status: CompactStatus,
     compact_task_assignment: BTreeMap<u64, CompactTaskAssignment>,
+    // TODO: may want fine-grained lock for each compaction group
```
Comment (soundOfDestiny) [inline src/meta/src/hummock/hummock_manager.rs:78]: if we already have different compact status for different group here, why do we need to filter in get_compact_task?

Comment (zwang28) [reply src/meta/src/hummock/hummock_manager.rs:78]: `compact_status` doesn't store the LSMT aka `levels` .
The `levels` passed into `get_compact_task` is a global one. We can filter it before passing it to `get_compact_task` of course.

Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs:78]: > `compact_status` doesn't store the LSMT aka `levels` .
> The `levels` passed into `get_compact_task` is a global one. We can filter it before passing it to `get_compact_task` of course.

sorry. I forgot that it has been refactored.

**Review by soundOfDestiny — APPROVED** [review]
lgtm after discussion with @zwang28

File: src/meta/src/hummock/compaction/level_selector.rs [hunk 3]
```diff
@@ -69,10 +67,7 @@ pub struct DynamicLevelSelector {
 impl Default for DynamicLevelSelector {
     fn default() -> Self {
         let config = Arc::new(CompactionConfig::default());
```
Comment (Little-Wallace) [inline src/meta/src/hummock/compaction/level_selector.rs:69]: Why not store `ProstCompactionConfig` directly

File: src/meta/src/hummock/hummock_manager.rs [hunk 4]
```diff
@@ -504,10 +494,18 @@ where
...
+        let mut compact_status = VarTransaction::new(
+            compaction
+                .compaction_statuses
+                .get_mut(&compaction_group_id)
+                .ok_or(Error::InvalidCompactionGroup(compaction_group_id))?,
+        );
         let current_version = self.versioning.read().await.current_version();
-        let compact_task = compact_status
-            .get_compact_task(&current_version.levels, task_id as HummockCompactionTaskId);
+        let compact_task = compact_status.get_compact_task(
```
Comment (Little-Wallace) [inline src/meta/src/hummock/hummock_manager.rs:504]: I thought the levels of each group would be split and independent....
So how could we filter the files by group_id?

Comment (zwang28) [reply src/meta/src/hummock/hummock_manager.rs:504]: > I thought the levels of each group would be split and independent....

Yes logically they are.

> So how could we filter the files by group_id?

I was thinking about identify a file's group_id
- Either by its key range. It requires both the key range's left and right are valid key from this file.
- Or by a group_id field in its meta.

Let me consider splitting them physically as you mentioned.

File: src/meta/src/hummock/compaction/level_selector.rs (hunk 3 above)
Comment (zwang28) [reply src/meta/src/hummock/compaction/level_selector.rs:69]: It could be.

File: src/meta/src/hummock/hummock_manager.rs (hunk 4 above)
Comment (zwang28) [reply src/meta/src/hummock/hummock_manager.rs:504]: I think we need a way to identify a file's group anyway. For example, even if we physically split groups' levels, we need to know each new files' groups in `commit_epoch`. 
Therefore, I prefer to not split groups' levels physically. What do you think?

Comment (zwang28) [reply src/meta/src/hummock/hummock_manager.rs:564]: Let's split levels by compaction group id in `HummockVersion` in following PR.

File: src/meta/src/hummock/compaction/level_selector.rs (hunk 3 above)
Comment (zwang28) [reply src/meta/src/hummock/compaction/level_selector.rs:69]: It's not convenient to add new methods for ProstCompactionConfig, so I'd like to keep CompactionConfig.

**Review by hzxa21 — APPROVED** [review]
LGTM.

File: src/meta/src/hummock/compaction/level_selector.rs (hunk 3 above)
Comment (Little-Wallace) [reply src/meta/src/hummock/compaction/level_selector.rs:69]: No.... Config is a lightweight structure， we do no need add too many methods for it

Comment (zwang28) [reply src/meta/src/hummock/compaction/level_selector.rs]: Fixed.
